// @alfresco/adf-core/esm5/notification-history/notification-history.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input, ViewChild } from '@angular/core';
import { NotificationService } from '../services/notification.service';
import { MatMenuTrigger } from '@angular/material';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
var NotificationHistoryComponent = /** @class */ (function () {
    function NotificationHistoryComponent(notificationService) {
        var _this = this;
        this.notificationService = notificationService;
        this.onDestroy$ = new Subject();
        this.notifications = [];
        /**
         * Custom choice for opening the menu at the bottom. Can be `before` or `after`.
         */
        this.menuPositionX = 'after';
        /**
         * Custom choice for opening the menu at the bottom. Can be `above` or `below`.
         */
        this.menuPositionY = 'below';
        this.notificationService.messages
            .pipe(takeUntil(this.onDestroy$))
            .subscribe((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            _this.notifications.push(message);
        }));
    }
    /**
     * @return {?}
     */
    NotificationHistoryComponent.prototype.isEmptyNotification = /**
     * @return {?}
     */
    function () {
        return (!this.notifications || this.notifications.length === 0);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NotificationHistoryComponent.prototype.onKeyPress = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.closeUserModal(event);
    };
    /**
     * @return {?}
     */
    NotificationHistoryComponent.prototype.markAsRead = /**
     * @return {?}
     */
    function () {
        this.notifications = [];
    };
    /**
     * @return {?}
     */
    NotificationHistoryComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.onDestroy$.next(true);
        this.onDestroy$.complete();
    };
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    NotificationHistoryComponent.prototype.closeUserModal = /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if ($event.keyCode === 27) {
            this.trigger.closeMenu();
        }
    };
    NotificationHistoryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-notification-history',
                    template: "<div (keyup)=\"onKeyPress($event)\">\n    <button mat-button [matMenuTriggerFor]=\"menu\" class=\"adf-notification-history-menu_button\"\n            id=\"adf-notification-history-open-button\">\n        <mat-icon>mail_outline</mat-icon>\n    </button>\n    <mat-menu #menu=\"matMenu\" [xPosition]=\"menuPositionX\" [yPosition]=\"menuPositionY\"\n              [overlapTrigger]=\"false\" id=\"adf-notification-history-menu\" class=\"adf-notification-history-menu\">\n\n        <div id=\"adf-notification-history-list\">\n            <mat-list>\n                <mat-list-item>\n                    <h6 mat-line>{{ 'NOTIFICATION_HISTORY.NOTIFICATIONS' | translate }}</h6>\n                </mat-list-item>\n            </mat-list>\n            <mat-divider></mat-divider>\n\n            <mat-list>\n                <mat-list-item *ngFor=\"let notification of notifications\">\n                    <mat-icon mat-list-icon>{{notification.info? notification.info: 'info'}}</mat-icon>\n                    <h4 mat-line>{{notification.message}}</h4>\n                    <p mat-line> {{notification.dateTime | date}} </p>\n                </mat-list-item>\n                <mat-list-item *ngIf=\"isEmptyNotification()\" id=\"adf-notification-history-component-no-message\">\n                    <h4 mat-line>{{ 'NOTIFICATION_HISTORY.NO_MESSAGE' | translate }}</h4>\n                </mat-list-item>\n                <mat-action-list *ngIf=\"!isEmptyNotification()\" id=\"adf-notification-history-mark-as-read\">\n                    <button mat-list-item (click)=\"markAsRead()\">{{ 'NOTIFICATION_HISTORY.MARK_AS_READ' | translate }}\n                    </button>\n                </mat-action-list>\n            </mat-list>\n        </div>\n    </mat-menu>\n</div>\n",
                    styles: [".adf-notification-history-menu_button.mat-button{margin-right:0;border-radius:90%;padding:0;min-width:40px;height:40px}@media only screen and (min-device-width:480px){.mat-menu-panel.adf-notification-history-menu{max-height:450px;min-width:450px;overflow:auto;padding:0}}.mat-menu-panel.adf-notification-history-menu .mat-menu-content{padding:0}"]
                }] }
    ];
    /** @nocollapse */
    NotificationHistoryComponent.ctorParameters = function () { return [
        { type: NotificationService }
    ]; };
    NotificationHistoryComponent.propDecorators = {
        trigger: [{ type: ViewChild, args: [MatMenuTrigger,] }],
        menuPositionX: [{ type: Input }],
        menuPositionY: [{ type: Input }]
    };
    return NotificationHistoryComponent;
}());
export { NotificationHistoryComponent };
if (false) {
    /** @type {?} */
    NotificationHistoryComponent.prototype.onDestroy$;
    /** @type {?} */
    NotificationHistoryComponent.prototype.notifications;
    /** @type {?} */
    NotificationHistoryComponent.prototype.trigger;
    /**
     * Custom choice for opening the menu at the bottom. Can be `before` or `after`.
     * @type {?}
     */
    NotificationHistoryComponent.prototype.menuPositionX;
    /**
     * Custom choice for opening the menu at the bottom. Can be `above` or `below`.
     * @type {?}
     */
    NotificationHistoryComponent.prototype.menuPositionY;
    /**
     * @type {?}
     * @private
     */
    NotificationHistoryComponent.prototype.notificationService;
}
