// @alfresco/adf-core/esm5/services/auth-guard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { AppConfigService } from '../app-config/app-config.service';
import { AuthGuardBase } from './auth-guard-base';
import { JwtHelperService } from './jwt-helper.service';
import * as i0 from "@angular/core";
import * as i1 from "./jwt-helper.service";
import * as i2 from "./authentication.service";
import * as i3 from "@angular/router";
import * as i4 from "../app-config/app-config.service";
var AuthGuard = /** @class */ (function (_super) {
    tslib_1.__extends(AuthGuard, _super);
    function AuthGuard(jwtHelperService, authenticationService, router, appConfigService) {
        var _this = _super.call(this, authenticationService, router, appConfigService) || this;
        _this.jwtHelperService = jwtHelperService;
        _this.ticketChangeBind = _this.ticketChange.bind(_this);
        window.addEventListener('storage', _this.ticketChangeBind);
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    AuthGuard.prototype.ticketChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.key === 'ticket-ECM' && event.newValue !== event.oldValue) {
            this.ticketChangeRedirect(event, 'ECM');
        }
        if (event.key === 'ticket-BPM' && event.newValue !== event.oldValue) {
            this.ticketChangeRedirect(event, 'BPM');
        }
        if (event.key === JwtHelperService.USER_ACCESS_TOKEN &&
            this.jwtHelperService.getValueFromToken(event.newValue, JwtHelperService.USER_PREFERRED_USERNAME) !==
                this.jwtHelperService.getValueFromToken(event.oldValue, JwtHelperService.USER_PREFERRED_USERNAME)) {
            this.ticketChangeRedirect(event, 'ALL');
        }
    };
    /**
     * @private
     * @param {?} event
     * @param {?} provider
     * @return {?}
     */
    AuthGuard.prototype.ticketChangeRedirect = /**
     * @private
     * @param {?} event
     * @param {?} provider
     * @return {?}
     */
    function (event, provider) {
        if (!event.newValue) {
            this.redirectToUrl(provider, this.router.url);
        }
        else {
            window.location.reload();
        }
        window.removeEventListener('storage', this.ticketChangeBind);
    };
    /**
     * @param {?} activeRoute
     * @param {?} redirectUrl
     * @return {?}
     */
    AuthGuard.prototype.checkLogin = /**
     * @param {?} activeRoute
     * @param {?} redirectUrl
     * @return {?}
     */
    function (activeRoute, redirectUrl) {
        if (this.authenticationService.isLoggedIn() || this.withCredentials) {
            return true;
        }
        if (!this.authenticationService.isOauth() || this.isOAuthWithoutSilentLogin()) {
            this.redirectToUrl('ALL', redirectUrl);
        }
        return false;
    };
    AuthGuard.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AuthGuard.ctorParameters = function () { return [
        { type: JwtHelperService },
        { type: AuthenticationService },
        { type: Router },
        { type: AppConfigService }
    ]; };
    /** @nocollapse */ AuthGuard.ngInjectableDef = i0.defineInjectable({ factory: function AuthGuard_Factory() { return new AuthGuard(i0.inject(i1.JwtHelperService), i0.inject(i2.AuthenticationService), i0.inject(i3.Router), i0.inject(i4.AppConfigService)); }, token: AuthGuard, providedIn: "root" });
    return AuthGuard;
}(AuthGuardBase));
export { AuthGuard };
if (false) {
    /** @type {?} */
    AuthGuard.prototype.ticketChangeBind;
    /**
     * @type {?}
     * @private
     */
    AuthGuard.prototype.jwtHelperService;
}
