// @alfresco/adf-core/esm5/services/automation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { AppConfigService } from '../app-config/app-config.service';
import { AlfrescoApiService } from '../services/alfresco-api.service';
import { StorageService } from './storage.service';
import { UserPreferencesService } from './user-preferences.service';
import { DemoForm } from '../mock/form/demo-form.mock';
import * as i0 from "@angular/core";
import * as i1 from "../app-config/app-config.service";
import * as i2 from "./alfresco-api.service";
import * as i3 from "./user-preferences.service";
import * as i4 from "./storage.service";
var CoreAutomationService = /** @class */ (function () {
    function CoreAutomationService(appConfigService, alfrescoApiService, userPreferencesService, storageService) {
        this.appConfigService = appConfigService;
        this.alfrescoApiService = alfrescoApiService;
        this.userPreferencesService = userPreferencesService;
        this.storageService = storageService;
        this.forms = new DemoForm();
    }
    /**
     * @return {?}
     */
    CoreAutomationService.prototype.setup = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var adfProxy = window['adf'] || {};
        adfProxy.setConfigField = (/**
         * @param {?} field
         * @param {?} value
         * @return {?}
         */
        function (field, value) {
            _this.appConfigService.config[field] = JSON.parse(value);
        });
        adfProxy.setStorageItem = (/**
         * @param {?} key
         * @param {?} data
         * @return {?}
         */
        function (key, data) {
            _this.storageService.setItem(key, data);
        });
        adfProxy.setUserPreference = (/**
         * @param {?} key
         * @param {?} data
         * @return {?}
         */
        function (key, data) {
            _this.userPreferencesService.set(key, data);
        });
        adfProxy.setFormInEditor = (/**
         * @param {?} json
         * @return {?}
         */
        function (json) {
            _this.forms.formDefinition = JSON.parse(json);
        });
        adfProxy.setCloudFormInEditor = (/**
         * @param {?} json
         * @return {?}
         */
        function (json) {
            _this.forms.cloudFormDefinition = JSON.parse(json);
        });
        adfProxy.clearStorage = (/**
         * @return {?}
         */
        function () {
            _this.storageService.clear();
        });
        adfProxy.apiReset = (/**
         * @return {?}
         */
        function () {
            _this.alfrescoApiService.reset();
        });
        window['adf'] = adfProxy;
    };
    CoreAutomationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CoreAutomationService.ctorParameters = function () { return [
        { type: AppConfigService },
        { type: AlfrescoApiService },
        { type: UserPreferencesService },
        { type: StorageService }
    ]; };
    /** @nocollapse */ CoreAutomationService.ngInjectableDef = i0.defineInjectable({ factory: function CoreAutomationService_Factory() { return new CoreAutomationService(i0.inject(i1.AppConfigService), i0.inject(i2.AlfrescoApiService), i0.inject(i3.UserPreferencesService), i0.inject(i4.StorageService)); }, token: CoreAutomationService, providedIn: "root" });
    return CoreAutomationService;
}());
export { CoreAutomationService };
if (false) {
    /** @type {?} */
    CoreAutomationService.prototype.forms;
    /**
     * @type {?}
     * @private
     */
    CoreAutomationService.prototype.appConfigService;
    /**
     * @type {?}
     * @private
     */
    CoreAutomationService.prototype.alfrescoApiService;
    /**
     * @type {?}
     * @private
     */
    CoreAutomationService.prototype.userPreferencesService;
    /**
     * @type {?}
     * @private
     */
    CoreAutomationService.prototype.storageService;
}
