// @alfresco/adf-core/esm5/services/directionality-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable, Inject, RendererFactory2 } from '@angular/core';
import { Directionality } from '@angular/cdk/bidi';
import { UserPreferencesService } from '../services/user-preferences.service';
import { DOCUMENT } from '@angular/platform-browser';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
import * as i2 from "./user-preferences.service";
import * as i3 from "@angular/cdk/bidi";
var DirectionalityConfigService = /** @class */ (function () {
    function DirectionalityConfigService(document, rendererFactory, userPreferencesService, directionality) {
        var _this = this;
        this.document = document;
        this.rendererFactory = rendererFactory;
        this.userPreferencesService = userPreferencesService;
        this.directionality = directionality;
        /** @type {?} */
        var renderer = this.rendererFactory.createRenderer(null, null);
        this.userPreferencesService
            .select('textOrientation')
            .subscribe((/**
         * @param {?} direction
         * @return {?}
         */
        function (direction) {
            renderer.setAttribute(_this.document.body, 'dir', direction);
            ((/** @type {?} */ (_this.directionality))).value = direction;
        }));
    }
    DirectionalityConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DirectionalityConfigService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: RendererFactory2 },
        { type: UserPreferencesService },
        { type: Directionality }
    ]; };
    /** @nocollapse */ DirectionalityConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DirectionalityConfigService_Factory() { return new DirectionalityConfigService(i0.inject(i1.DOCUMENT), i0.inject(i0.RendererFactory2), i0.inject(i2.UserPreferencesService), i0.inject(i3.Directionality)); }, token: DirectionalityConfigService, providedIn: "root" });
    return DirectionalityConfigService;
}());
export { DirectionalityConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DirectionalityConfigService.prototype.document;
    /**
     * @type {?}
     * @private
     */
    DirectionalityConfigService.prototype.rendererFactory;
    /**
     * @type {?}
     * @private
     */
    DirectionalityConfigService.prototype.userPreferencesService;
    /**
     * @type {?}
     * @private
     */
    DirectionalityConfigService.prototype.directionality;
}
