// @alfresco/adf-core/esm5/services/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export { AuthGuardBase } from './auth-guard-base';
export { AuthenticationService } from './authentication.service';
export { AlfrescoApiService } from './alfresco-api.service';
export { ContentService } from './content.service';
export { AuthGuard } from './auth-guard.service';
export { AuthGuardEcm } from './auth-guard-ecm.service';
export { AuthGuardBpm } from './auth-guard-bpm.service';
export { AuthGuardSsoRoleService } from './auth-guard-sso-role.service';
export { AppsProcessService } from './apps-process.service';
export { PageTitleService } from './page-title.service';
export { StorageService } from './storage.service';
export { CookieService } from './cookie.service';
export { RenditionsService } from './renditions.service';
export { NotificationService } from './notification.service';
export { LogService } from './log.service';
export { TRANSLATION_PROVIDER, TranslationService } from './translation.service';
export { TranslateLoaderService } from './translate-loader.service';
export { ThumbnailService } from './thumbnail.service';
export { UploadService } from './upload.service';
export { DynamicComponentResolver, DynamicComponentMapper } from './dynamic-component-mapper.service';
export { UserPreferenceValues, UserPreferencesService } from './user-preferences.service';
export { HighlightTransformService } from './highlight-transform.service';
export { DeletedNodesApiService } from './deleted-nodes-api.service';
export { FavoritesApiService } from './favorites-api.service';
export { NodesApiService } from './nodes-api.service';
export { PeopleContentService } from './people-content.service';
export { PeopleProcessService } from './people-process.service';
export { SearchService } from './search.service';
export { SharedLinksApiService } from './shared-links-api.service';
export { SitesService } from './sites.service';
export { DiscoveryApiService } from './discovery-api.service';
export { CommentProcessService } from './comment-process.service';
export { SearchConfigurationService } from './search-configuration.service';
export { CommentContentService } from './comment-content.service';
export { LoginDialogService } from './login-dialog.service';
export { ExternalAlfrescoApiService } from './external-alfresco-api.service';
export { JwtHelperService } from './jwt-helper.service';
export { DownloadZipService } from './download-zip.service';
export { LockService } from './lock.service';
export { CoreAutomationService } from './automation.service';
export { PreviousRouteService } from './previous-route.service';
export {} from './automation.service';
